typed-racket-lib/typed-racket/base-env/unit-prims.rkt

```diff
@@ -0,0 +1,767 @@
+           #t)))])
+
+(define-syntax (define-compound-unit stx)
+  (syntax-parse stx
+    [(_ uid 
+        imports:compound-imports
+        exports:compound-exports
+        links:compound-links)
+     (quasisyntax/loc stx
+       (process-define-compound-unit
+        (void #,@#'imports.import-link-ids #,@(attribute links.bound-link-ids))
+        (void #,@#'imports.import-sig-ids #,@(attribute links.bound-sig-ids))
+        (void #,@#'imports.import-link-ids)
+        (void)
+        exports.export-link-ids
+        (untyped-define-compound-unit uid
+                                      imports
+                                      exports
+                                      links.untyped-links)))]))
+
+;; compound-unit/infer
+(begin-for-syntax
+  (define-syntax-class compound-infer-imports
+    #:literals (import)
+    (pattern (import im:infer-link-import ...)
+             #:with import-link-ids
+             #'(im.link-qs ...)
+             #:with import-sig-ids
+             #'(im.sig-qs ...)))
+  
+  (define-syntax-class compound-infer-exports
+    #:literals (export)
+    (pattern (export ex:infer-link-export ...)
+             #:with export-links-or-sigs
+             #'(void (quote-syntax ex.link-or-sig-id) ...)))
+  
+  (define-syntax-class compound-infer-links
+    #:literals (link)
+    (pattern (link lnk:infer-linkage-decl ...)
+             ;; #:with untyped-links
+             ;; #'(link lnk.untyped-link-decl ...)
+             #:attr bound-link-ids (apply append (map syntax->list 
+                                                      (syntax->list 
+                                                       #'(lnk.bound-link-ids ...))))
+             #:attr bound-sig-ids (apply append (map syntax->list
+                                                     (syntax->list
+                                                      #'(lnk.bound-sig-ids ...))))
+             #:with links-untyped 
+             #'(link lnk.linkage-stx ...)
+             #:attr link-table
+             #`(void lnk.linkage-information ...)))
+  
+  (define-syntax-class infer-link-import
+    (pattern sig-id:id
+             #:with sig-qs #'(quote-syntax sig-id)
+             #:with link-qs #`(quote-syntax #,(generate-temporary)))
+    (pattern (link-id:id : sig-id:id)
+             #:with link-qs #'(quote-syntax link-id)
+             #:with sig-qs #'(quote-syntax sig-id)))
+  
+  (define-syntax-class infer-link-export
+    (pattern link-or-sig-id:id))
+  
+  (define-syntax-class infer-linkage-decl
+    (pattern ((lb:link-binding ...)
+              unit-id:id
+              link-id:id ...)
+             #:with bound-link-ids #'(lb.link-qs ...)
+             #:with bound-sig-ids #'(lb.sig-qs ...)
+             #:with linkage-stx
+             #`((lb ...)
+                #,(tr:unit:compound:expr-property
+                   #'unit-id
+                   'infer)
+                link-id ...)
+             #:with linkage-information
+             (begin
```

Comment:
This attribute expression looks exactly the same as the one below. Abstract with a helper function.